fwprovider/resource_download_file.go

```diff
@@ -0,0 +1,446 @@
+				},
+			},
+			"checksum": schema.StringAttribute{
+				MarkdownDescription: "The expected checksum of the file.",
+				Optional:            true,
+				Default:             nil,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+				Validators: []validator.String{
+					stringvalidator.AlsoRequires(path.MatchRoot("checksum_algorithm")),
+				},
+			},
+			"compression": schema.StringAttribute{
+				MarkdownDescription: "Decompress the downloaded file using the " +
+					"specified compression algorithm.",
+				Optional: true,
+				Default:  nil,
+			},
+			"checksum_algorithm": schema.StringAttribute{
+				MarkdownDescription: "The algorithm to calculate the checksum of the file. " +
+					"Must be `md5` | `sha1` | `sha224` | `sha256` | `sha384` | `sha512`.",
+				Optional: true,
+				Validators: []validator.String{
+					stringvalidator.OneOf([]string{
+						"md5",
+						"sha1",
+						"sha224",
+						"sha256",
+						"sha384",
+						"sha512",
+					}...),
+					stringvalidator.AlsoRequires(path.MatchRoot("checksum")),
+				},
+				Default: nil,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"verify": schema.BoolAttribute{
+				MarkdownDescription: "By default `true`. If `false`, no SSL/TLS certificates will be verified.",
+				Optional:            true,
+				Computed:            true,
+				Default:             booldefault.StaticBool(true),
+			},
+			"allow_unsupported_types": schema.BoolAttribute{
+				MarkdownDescription: "By default `false`. If `true`, " +
+					"content formats `qcow2` and `raw` can be downloaded, " +
+					"though it is not supported by proxmox.",
+				Optional: true,
+				Computed: true,
+				Default:  booldefault.StaticBool(false),
+			},
+		},
+	}
+}
+
+func (r *downloadFileResource) Configure(
+	_ context.Context,
+	req resource.ConfigureRequest,
+	resp *resource.ConfigureResponse,
+) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(proxmox.Client)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *proxmox.Client, got: %T", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.client = client
+}
+
+func (r *downloadFileResource) Create(
+	ctx context.Context,
+	req resource.CreateRequest,
+	resp *resource.CreateResponse,
+) {
+	var plan downloadFileModel
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	nodesClient := r.client.Node(plan.Node.ValueString())
+	verify := proxmoxtypes.CustomBool(plan.Verify.ValueBool())
+
+	queryURLMetadataReq := nodes.QueryURLMetadataGetRequestBody{
+		URL:    plan.URL.ValueStringPointer(),
+		Verify: &verify,
+	}
+
+	fileMetadata, err := nodesClient.GetQueryURLMetadata(
+		ctx,
+		&queryURLMetadataReq,
+	)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating Download File interface",
```

Comment:
TBH, I was really confused with the error message when saw it first time 😅 
```suggestion
			"Error initiating file download",
```

Reply:
done